am/src/commands/instrument.rs

```diff
@@ -0,0 +1,115 @@
+    /// - For Rust projects it must be where the Cargo.toml lie,
+    /// - For Go projects it must be the root of the repository,
+    /// - For Python projects it must be the root of the library,
+    /// - For Typescript projects it must be where the package.json lie.
+    #[arg(value_name = "ROOT", verbatim_doc_comment)]
+    root: PathBuf,
+    /// A list of patterns to exclude from instrumentation. The patterns follow .gitignore rules, so
+    /// `--exclude "/vendor/"` will exclude all the vendor subdirectory only at the root, and adding
+    /// a pattern that starts with `!` will unignore a file or directory
+    #[arg(short, long, value_name = "PATTERNS")]
+    exclude: Vec<String>,
+}
+
+#[derive(Args)]
+struct AllProjects {
+    /// Main directory to start the subprojects search on. am currently detects
+    /// Rust (Cargo.toml), Typescript (package.json), and Golang (go.mod)
+    /// projects.
+    #[arg(value_name = "ROOT")]
+    root: PathBuf,
+    /// A list of patterns to exclude from instrumentation. The patterns follow .gitignore rules, so
+    /// `--exclude "/vendor/"` will exclude all the vendor subdirectory only at the root, and adding
+    /// a pattern that starts with `!` will unignore a file or directory
+    #[arg(short, long, value_name = "PATTERNS")]
+    exclude: Vec<String>,
+}
+
+pub fn handle_command(args: Arguments) -> anyhow::Result<()> {
+    match args.command {
+        Command::Single(args) => handle_single_project(args),
+        Command::All(args) => handle_all_projects(args),
+    }
+}
+
+fn handle_all_projects(args: AllProjects) -> Result<(), anyhow::Error> {
+    let root = args
+        .root
+        .canonicalize()
+        .context("The path must be resolvable to an absolute path")?;
+    info!("Instrumenting functions in {}:", root.display());
+
+    let mut exclude_patterns_builder = ignore::gitignore::GitignoreBuilder::new(&root);
+    for pattern in args.exclude {
+        exclude_patterns_builder.add_line(None, &pattern)?;
+    }
+    let exclude_patterns = exclude_patterns_builder.build()?;
+
+    am_list::instrument_all_project_files(&root, &exclude_patterns)?;
+
+    println!("If your project has Golang files, you need to run `go generate` now.");
+
+    Ok(())
+}
+
+fn handle_single_project(args: SingleProject) -> Result<(), anyhow::Error> {
+    let root = args
+        .root
+        .canonicalize()
+        .context("The path must be resolvable to an absolute path")?;
+    info!("Instrumenting functions in {}:", root.display());
+
+    let mut exclude_patterns_builder = ignore::gitignore::GitignoreBuilder::new(&root);
+    for pattern in args.exclude {
+        exclude_patterns_builder.add_line(None, &pattern)?;
+    }
+    let exclude_patterns = exclude_patterns_builder.build()?;
+
+    am_list::instrument_single_project_files(&root, args.language, &exclude_patterns)?;
+
+    if args.language == Language::Go {
+        println!("You need to run `go generate` now.");
```

Comment:
we cannot do this automatically? or is that a bad idea

Reply:
I don't know if it's common to have multiple go versions (with different names than `go`), and I don't have a Windows machine to test non-unix environments, so I'd rather leave it as an exercise. But usually IDEs/LSP clients will have a way to detect and run `go generate` easily